daemon/cluster/cluster.go

```diff
@@ -625,7 +605,16 @@ func (c *Cluster) Update(version uint64, spec types.Spec, flags types.UpdateFlag
 		return err
 	}
 
-	swarmSpec, err := convert.SwarmSpecToGRPC(spec)
+	// In update, client should specify the desired state, which includes
+	// Name and Labels. So we need to initialize the name and labels here
+	// based on the value in spec.
+	// It is possible that empty Labels be specified (unset in update).
+	swarmSpec, err := convert.SwarmSpecToGRPC(spec, swarmapi.ClusterSpec{
+		Annotations: swarmapi.Annotations{
+			Name:   spec.Name,
+			Labels: spec.Labels,
+		},
+	})
```

Comment:
> Thanks @aaronlehmann. At the client side (api/client/swarm/update.go) the existing swarm setting was obtained and then merged with the setting updates. So the updated setting is reflected in spec here (without getSwarm).

That makes sense.

> But it is better to obtain the existing swarm setting in daemon side here (with getSwarm and then update as needed) as end user may not use client directly (vs. remote API). Processing at the daemon side will prevent incorrect settings through remote API.

I'm actually now thinking that it's better not to do this at the API level for `swarm update`. If a user sets a value to 0 through the API, they would expect it to get reset to the default, but what happens instead is that the value stays unchanged. We don't have a way to fetch manager-side default values except when we are creating the cluster, so it's difficult to support resetting fields to the defaults.

So I think it's probably better to require a complete spec for `swarm update`, and return an error if anything that shouldn't be `0`/`nil` is set to `0` or `nil`. This also keeps engine-api more consistent with the GRPC api, where the client is expected to always form the spec on its own.

Sorry for the misdirection - it just took me a few passes to wrap my mind around this.